interfaces/cython/cantera/base.pyx

```diff
@@ -3,29 +3,27 @@
 
 cdef class _SolutionBase:
     def __cinit__(self, infile='', phaseid='', phases=(), origin=None,
-                  source=None, thermo=None, species=(), kinetics=None,
-                  reactions=(), **kwargs):
+                  source=None, yaml=None, thermo=None, species=(),
+                  kinetics=None, reactions=(), **kwargs):
         # Shallow copy of an existing Solution (for slicing support)
         cdef _SolutionBase other
         if origin is not None:
             other = <_SolutionBase?>origin
 
-            # keep a reference to the parent to prevent the underlying
-            # C++ objects from being deleted
-            self.parent = origin
-            self.is_slice = True
-
             self.thermo = other.thermo
             self.kinetics = other.kinetics
             self.transport = other.transport
+            self._thermo = other._thermo
+            self._kinetics = other._kinetics
+            self._transport = other._transport
 
             self.thermo_basis = other.thermo_basis
             self._selected_species = other._selected_species.copy()
             return
 
-        self.is_slice = False
-
-        if infile or source:
+        if infile.endswith('.yml') or infile.endswith('.yaml') or yaml:
```

Comment:
I think the suggested approach would make seeing error messages in either case more difficult to read, since you'd always end up with two error messages and two tracebacks.

What we can do, once the CTI/XML formats are removed, is let the `source` argument be interpreted as a YAML string as well, and just require that only one of the two is given.

Reply:
> I think the suggested approach would make seeing error messages in either case more difficult to read, since you'd always end up with two error messages and two tracebacks.

You could avoid the double traceback by

```python
try:
    self._init_yaml(...)
except:
    init = False
else:
    init = True

if not init:
    self._init_cti_xml(...)
```

The problem with this approach (with the bare `except`) is that if the YAML source is valid YAML syntax but has some problems with the data (e.g., missing parameters for a thermo model), that's treated the same as invalid YAML. Then, the CTI/XML is tried, but that's not going to be able to report that there's an error with the data. If there's a specific exception that's raised when the YAML is not able to be parsed, that should be caught directly instead of catching everything. Is the exception that `yaml-cpp` raises when the YAML is invalid different from one that says something is wrong with the data in the file? Is it able to be sent up to the Python interface?

> What we can do, once the CTI/XML formats are removed, is let the `source` argument be interpreted as a YAML string as well, and just require that only one of the two is given.

This will result in one more deprecation warning we have to manage when we want to remove the `yaml` argument, which I think we should once CTI/XML are gone. I guess that might be easier than trying to make sure the right exception is passed through from C++ into Python here.

Reply:
Yes, there are specific errors raised by the YAML parser on syntax errors, but I don't think you can reliably distinguish between "this is a YAML file with a syntax error" (where you want to display that error and whatever context you can provide) and "this isn't even supposed to be a YAML file". So, my inclination is to go with the extra argument, and (maybe) deprecate the new argument in a later release.